File: sample/kotlinapp/src/main/java/com/marcinmoskala/kotlinapp/MainActivity.java
package com.marcinmoskala.kotlinapp;

import android.os.Bundle;
import android.support.design.widget.TextInputLayout;
import android.view.View;
import android.widget.AutoCompleteTextView;
import android.widget.Button;
import android.widget.Switch;

public class MainActivity extends BaseActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Button showDataButton = (Button) findViewById(R.id.show_data_button);
        Button showParcelableDataButton = (Button) findViewById(R.id.show_parcelable_data_button);
        Button showSerializableDataButton = (Button) findViewById(R.id.show_serializable_data_button);
        final AutoCompleteTextView studentNameView = (AutoCompleteTextView) findViewById(R.id.student_name);
        final AutoCompleteTextView studentIdView = (AutoCompleteTextView) findViewById(R.id.student_id);
        final AutoCompleteTextView studentGradeView = (AutoCompleteTextView) findViewById(R.id.student_grade);
        final TextInputLayout studentGradeLayoutView = (TextInputLayout) findViewById(R.id.student_grade_layout);
        final Switch studentIsPassingView = (Switch) findViewById(R.id.student_is_passing);

        showDataButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String gradeString = studentGradeView.getText().toString();
                if(gradeString.length() != 1) {
                    studentGradeLayoutView.setError("You must provide some grade");
                    return;
                } else {
                    studentGradeLayoutView.setError(null);
                }

                String name = studentNameView.getText().toString();
                String idString = studentIdView.getText().toString();
                char grade = gradeString.charAt(0);
                boolean isPassing = studentIsPassingView.isChecked();

                try {
                    int id = Integer.parseInt(idString);
                    if(name.trim().equals("")) {
                        StudentDataActivityStarter.start(getBaseContext(), id, grade, isPassing);
                    } else {
                        StudentDataActivityStarter.start(getBaseContext(), name, id, grade, isPassing);
                    }
                } catch (NumberFormatException e) {
                    // Id is not valid
                    if(name.trim().equals("")) {
                        StudentDataActivityStarter.start(getBaseContext(), grade, isPassing);
                    } else {
                        StudentDataActivityStarter.start(getBaseContext(), name, grade, isPassing);
                    }
                }
            }
        });
        showParcelableDataButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                StudentParcelable student = new StudentParcelable(10, "Marcin", 'A');
                StudentParcelableActivityStarter.start(getBaseContext(), student);
            }
        });
        showSerializableDataButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                StudentSerializable student = new StudentSerializable(20, "Marcin Moskala", 'A', true);
                StudentSerializableActivityStarter.start(getBaseContext(), student);
            }
        });
    }
}

File: sample/kotlinapp/src/main/java/com/marcinmoskala/kotlinapp/MainActivity.kt
package com.marcinmoskala.kotlinapp

import activitystarter.MakeActivityStarter
import android.app.AlarmManager.RTC_WAKEUP
import android.app.PendingIntent
import android.app.PendingIntent.FLAG_UPDATE_CURRENT
import android.os.Bundle
import com.marcinmoskala.kotlinapp.notification.NotificationPublisherStarter
import kotlinx.android.synthetic.main.activity_main.*
import org.joda.time.DateTime

@MakeActivityStarter
class MainActivity : BaseActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        showDataButton.setOnClickListener { startDetailsActivity() }
        showParcelableDataButton.setOnClickListener { startParcelableActivity() }
        showSerializableDataButton.setOnClickListener { startSerializableActivity() }
        startNotification()
    }

    private fun startNotification() {
        val intent = NotificationPublisherStarter.getIntent(this, notificationId, notificationTime)
        val pendingIntent = PendingIntent.getBroadcast(this, 0, intent, FLAG_UPDATE_CURRENT)
        alarmManager.set(RTC_WAKEUP, System.currentTimeMillis() + 1000, pendingIntent)
    }

    fun startDetailsActivity() {
        val gradeString = studentGradeView.text.toString()
        if (gradeString.length != 1) {
            studentGradeLayoutView.error = "You must provide some grade"
            return
        } else {
            studentGradeLayoutView.error = null
        }

        val name = studentNameView.text.toString()
        val idString = studentIdView.text.toString()
        val grade = gradeString[0]
        val isPassing = studentIsPassingView.isChecked

        try {
            val id = idString.toInt()
            if (name.isNullOrBlank()) {
                StudentDataActivityStarter.start(this, id, grade, isPassing)
            } else {
                StudentDataActivityStarter.start(this, name, id, grade, isPassing)
            }
        } catch (e: NumberFormatException) {
            // Id is not valid
            if (name.isNullOrBlank()) {
                StudentDataActivityStarter.start(this, grade, isPassing)
            } else {
                StudentDataActivityStarter.start(this, name, grade, isPassing)
            }
        }
    }

    fun startParcelableActivity() {
        StudentParcelableActivityStarter.start(this, parcelableStudent)
    }

    fun startSerializableActivity() {
        StudentSerializableActivityStarter.start(this, serializableStudent)
    }

    companion object {
        val notificationId = 10
        val notificationTime = DateTime.now().toString("HH:mm")
        val parcelableStudent = StudentParcelable(10, "Marcin", 'A')
        val serializableStudent = StudentSerializable(20, "Marcin Moskala", 'A', true)
    }
}
